pocs/tests/utils/test_utils.py

```diff
@@ -115,3 +116,149 @@ def test_countdown_timer():
     assert timer.expired() is True
+
+
+def test_delay_of_sigterm_with_nosignal():
+    orig_sigterm_handler = signal.getsignal(signal.SIGTERM)
+
+    with DelaySigTerm():
+        assert signal.getsignal(signal.SIGTERM) != orig_sigterm_handler
+
+    assert signal.getsignal(signal.SIGTERM) == orig_sigterm_handler
+
+
+def test_delay_of_sigterm_with_handled_signal():
+    """Confirm that another type of signal can be handled.
+
+    In this test we'll send SIGCHLD, which should call the
+    handler the test installs immediately.
+    """
+    test_signal = signal.SIGCHLD
+
+    # Booleans to keep track of how far we've gotten.
+    before_signal = False
+    after_signal = False
+    signal_handled = False
+    after_with = False
+
+    def signal_handler(signum, frame):
+        assert before_signal
+
+        nonlocal signal_handled
+        assert not signal_handled
+        signal_handled = True
+
+        assert not after_signal
+
+    old_test_signal_handler = signal.getsignal(test_signal)
+    orig_sigterm_handler = signal.getsignal(signal.SIGTERM)
+    try:
+        # Install our handler.
+        signal.signal(test_signal, signal_handler)
+
+        with DelaySigTerm():
+            assert signal.getsignal(signal.SIGTERM) != orig_sigterm_handler
+            before_signal = True
+            # Send the test signal. It should immediately
+            # call our handler.
+            os.kill(os.getpid(), test_signal)
+            assert signal_handled
+            after_signal = True
+
+        after_with = True
+        assert signal.getsignal(signal.SIGTERM) == orig_sigterm_handler
+    finally:
+        assert before_signal
+        assert signal_handled
+        assert after_signal
+        assert after_with
+        assert signal.getsignal(signal.SIGTERM) == orig_sigterm_handler
+        signal.signal(test_signal, old_test_signal_handler)
+
+
+def test_delay_of_sigterm_with_raised_exception():
+    """Confirm that raising an exception inside the handler is OK."""
+    test_signal = signal.SIGCHLD
+
+    # Booleans to keep track of how far we've gotten.
+    before_signal = False
+    after_signal = False
+    signal_handled = False
+    exception_caught = False
+
+    def signal_handler(signum, frame):
+        assert before_signal
+
+        nonlocal signal_handled
+        assert not signal_handled
+        signal_handled = True
+
+        assert not after_signal
+        raise UserWarning()
+
+    old_test_signal_handler = signal.getsignal(test_signal)
+    orig_sigterm_handler = signal.getsignal(signal.SIGTERM)
+    try:
+        # Install our handler.
+        signal.signal(test_signal, signal_handler)
+
+        with DelaySigTerm():
+            assert signal.getsignal(signal.SIGTERM) != orig_sigterm_handler
+            before_signal = True
+            # Send the test signal. It should immediately
+            # call our handler.
+            os.kill(os.getpid(), test_signal)
+            after_signal = True
```

Comment:
How about:

```python
after_signal = True  # Should not get here
assert after_signal
```
just to make it clear that it shouldn't get to either of these statements.

Reply:
Added a bunch of comments explaining, rather than the proposed code.